DBoW2/DBoW2_sample_code.cpp

```diff
@@ -0,0 +1,218 @@
+{
+  EASY_FUNCTION("Load DB Features", profiler::colors::LightGreen);
+
+  vfeatures.clear();
+  vfeatures.reserve(NDBIMAGES);
+
+  for(fs::directory_iterator it(dbPath); it != fs::end(it); it++)
+  {
+    const fs::directory_entry &entry = *it;
+    std::string fileName = entry.path();
+    cv::Mat image = cv::imread(fileName, 0);
+    if(image.empty()) return false;
+    cv::Mat mask;
+    std::vector<cv::KeyPoint> keypoints;
+    cv::Mat descriptors;
+
+    orb->detectAndCompute(image, mask, keypoints, descriptors);
+
+    dbTable.push_back(fileName);
+    vfeatures.push_back(UnitTest::Features());
+    changeStructure(descriptors, vfeatures.back());
+  }
+
+  return true;
+  
+}
+
+
+void changeStructure(const cv::Mat &plain, UnitTest::Features &out)
+{
+  out.resize(plain.rows);
+
+  for(int i = 0; i < plain.rows; ++i)
+    out[i] = plain.row(i);
+  
+}
+
+
+bool loadVocabulary(OrbVocabulary &voc)
+{
+  EASY_FUNCTION("Load Vocabulary", profiler::colors::Yellow);
+  
+  if(!voc.loadFromTextFile(vocPath))  return false;
+
+  std::cout << "< Vocabulary information >" << std::endl
+  << voc << std::endl << std::endl;
+
+  return true;
+
+}
+
+
+bool createDatabase(OrbDatabase &db, const UnitTest::FeatureVector &vfeatures)
+{
+  EASY_FUNCTION("Create Database", profiler::colors::Magenta);
+
+  for(int i = 0; i < NDBIMAGES; i++)
+    db.add(vfeatures[i]);
+
+  std::cout << "< Database information >\n" << db << std::endl << std::endl;
+
+  return true;
+}
+
+
+bool loadQueryFeatures(UnitTest::FeatureVector &vfeatures, const cv::Ptr<cv::ORB> &orb)
+{
+  EASY_FUNCTION("Load Query Features", profiler::colors::DeepOrange);
+
+  vfeatures.clear();
+  vfeatures.reserve(NQUERYIMAGES);
+
+  // Feature Extraction for Query Images 
+  for(fs::directory_iterator it(queryPath); it != fs::end(it); it++)
+  {
+    const fs::directory_entry &entry = *it;
+    std::string fileName = entry.path();
+    cv::Mat image = cv::imread(fileName, 0);
+    if(image.empty()) return false;
+    cv::Mat mask;
+    std::vector<cv::KeyPoint> keypoints;
+    cv::Mat descriptors;
+
+    orb->detectAndCompute(image, mask, keypoints, descriptors);
+
+    vfeatures.push_back(UnitTest::Features());
+    changeStructure(descriptors, vfeatures.back());
+  }
+
+  return true;
+
+}
+
+
+bool queryDatabase(OrbDatabase &db, const UnitTest::FeatureVector &vfeatures, UnitTest::FileLUT &queryResultTable, const UnitTest::FileLUT &dbTable)
+{
+  EASY_FUNCTION("Query Database", profiler::colors::DarkTeal);
+
+  QueryResults ret;
+  int nCandidates = 1;
+  if(nCandidates > NQUERYIMAGES)  return false; 
+
+  for(int i = 0; i < NQUERYIMAGES; i++)
+  {
+    // Query & Save "nCandidates" best match in QueryResult
+    db.query(vfeatures[i], ret, nCandidates);
+
+    std::cout << "Searching for Image " << i << ". " << ret << std::endl << std::endl;
+
+    // Save Image index of the Best Match for each Query Image
+    queryResultTable.push_back(dbTable[ret[0].Id]);

```

Comment:
```suggestion
    const auto& bestMatch = dbTable[ret[0].Id]);
    queryResultTable.push_back(bestMatch);

```
Reference variable을 이용해서 가독성을 높일 수 있습니다.